### PR TITLE
Structures, Tools, Tasks, are now serializable.

### PR DESCRIPTION
**Pull request by william-price01**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
- Added `SerializableMixin` to `Structure`, `BaseTask`, `BaseTool`, `ToolTask`, `TaskMemory`.
- Added `metadata={"serializable": True}` to multiple fields on `Structure`, `BaseTask`, `BaseTool`, `ToolTask`, `TaskMemory`

### CODE REVIEW

File: griptape/tasks/base_task.py [hunk 1]
```diff
@@ -22,21 +23,21 @@
...
         EXECUTING = 2
         FINISHED = 3
 
-    id: str = field(default=Factory(lambda: uuid.uuid4().hex), kw_only=True)
-    state: State = field(default=State.PENDING, kw_only=True)
+    id: str = field(default=Factory(lambda: uuid.uuid4().hex), kw_only=True, metadata={"serializable": True})
+    state: State = field(default=State.PENDING, kw_only=True, metadata={"serializable": True})
     parent_ids: list[str] = field(factory=list, kw_only=True)
     child_ids: list[str] = field(factory=list, kw_only=True)
     max_meta_memory_entries: Optional[int] = field(default=20, kw_only=True)
```
Comment (collindutter) [inline griptape/tasks/base_task.py:36]: Let's serialize these

**Review by collindutter — COMMENTED** [review]
Awesome! I know CodeCov is passing but please add `test_to_dict` tests for each of the classes.

File: griptape/tasks/base_task.py (hunk 1 above)
Comment (william-price01) [reply griptape/tasks/base_task.py:36]: sounds good!

File: CHANGELOG.md [hunk 2]
```diff
@@ -38,6 +38,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `Chat` input now uses a slightly customized version of `Rich.prompt.Prompt` by default.
 - `Chat` output now uses `Rich.print` by default.
 - `Chat.output_fn`'s now takes an optional kwarg parameter, `stream`.
+- `Structure` now utilizes `SerializableMixin` for serialization.
+- `BaseTask` now utilizes `SerializableMixin` for serialization.
+- `BaseTool` now utilizes `SerializableMixin` for serialization.
+- `TaskMemory` now utilizes `SerializableMixin` for serialization.
```
[the remark this answers is not in the document]
Comment (william-price01) [reply CHANGELOG.md:44]: got ya.

**Review by collindutter — APPROVED** [review]
Fantastic work!

File: tests/unit/memory/tool/test_task_memory.py [hunk 3]
```diff
@@ -92,3 +92,27 @@ def test_load_artifacts_for_blob_list_artifact(self, memory):
...
+        deserialized_memory = memory.from_dict(serialized_memory)
+        assert isinstance(deserialized_memory, TaskMemory)
+
+        deserialized_memory.process_output(
+            MockTool().test,
+            ActionsSubtask(),
+            ListArtifact([BlobArtifact(b"foo", name="test1"), BlobArtifact(b"foo", name="test2")], name="test"),
+        )
+
+        assert len(deserialized_memory.load_artifacts("test")) == 2
```
Comment (collindutter) [inline tests/unit/memory/tool/test_task_memory.py:118]: Rename `test_to_dict` and `test_from_dict`

File: tests/unit/structures/test_agent.py [hunk 4]
```diff
@@ -252,3 +252,47 @@ def test_task_outputs(self):
...
+        assert isinstance(serialized_agent, dict)
+
+        deserialized_agent = Agent.from_dict(serialized_agent)
+        assert isinstance(deserialized_agent, Agent)
+
+        assert deserialized_agent.task_outputs[task.id] is None
+        deserialized_agent.run()
+
+        assert len(deserialized_agent.task_outputs) == 1
+        assert deserialized_agent.task_outputs[task.id].value == "mock output"
```
Comment (collindutter) [inline tests/unit/structures/test_agent.py:298]: Rename to `test_to_dict` and `test_from_dict`

File: tests/unit/structures/test_pipeline.py [hunk 5]
```diff
@@ -411,3 +411,47 @@ def test_task_outputs(self):
...
+        assert isinstance(serialized_pipeline, dict)
+
+        deserialized_pipeline = Pipeline.from_dict(serialized_pipeline)
+        assert isinstance(deserialized_pipeline, Pipeline)
+
+        assert deserialized_pipeline.task_outputs[task.id] is None
+        deserialized_pipeline.run()
+
+        assert len(deserialized_pipeline.task_outputs) == 1
+        assert deserialized_pipeline.task_outputs[task.id].value == "mock output"
```
Comment (collindutter) [inline tests/unit/structures/test_pipeline.py:457]: Rename to `test_to_dict` and `test_from_dict`

File: tests/unit/structures/test_workflow.py [hunk 6]
```diff
@@ -979,3 +979,47 @@ def test_task_outputs(self):
...
+        assert isinstance(serialized_workflow, dict)
+
+        deserialized_workflow = Workflow.from_dict(serialized_workflow)
+        assert isinstance(deserialized_workflow, Workflow)
+
+        assert deserialized_workflow.task_outputs[task.id] is None
+        deserialized_workflow.run()
+
+        assert len(deserialized_workflow.task_outputs) == 1
+        assert deserialized_workflow.task_outputs[task.id].value == "mock output"
```
Comment (collindutter) [inline tests/unit/structures/test_workflow.py:1025]: Rename to `test_to_dict` and `test_from_dict`

File: tests/unit/tasks/test_base_task.py [hunk 7]
```diff
@@ -120,7 +120,6 @@ def test_add_parent(self, task):
         agent = Agent()
         parent = MockTask("parent foobar", id="parent_foobar", structure=agent)
 
-        result = task.add_parent(parent)
```
Comment (collindutter) [inline tests/unit/tasks/test_base_task.py:123]: Why was this removed?

File: tests/unit/tasks/test_base_task.py [hunk 8]
```diff
@@ -159,3 +158,35 @@ def test_add_child_bitshift(self, task):
...
+        workflow = Workflow()
+        workflow.add_task(deserialized_task)
+
+        assert workflow.tasks == [deserialized_task]
+
+        workflow.run()
+
+        assert str(workflow.tasks[0].state) == "State.FINISHED"
+        assert workflow.tasks[0].id == deserialized_task.id
+        assert workflow.tasks[0].output.value == "foobar"
```
Comment (collindutter) [inline tests/unit/tasks/test_base_task.py:192]: Rename to `test_to_dict` and `test_from_dict`

File: tests/unit/tasks/test_tool_task.py [hunk 9]
```diff
@@ -237,3 +237,39 @@ def test_actions_schema(self):
...
+    def test_tool_task_from_dict(self):
+        tool = MockTool()
+        task = ToolTask("test", tool=tool)
+
+        serialized_tool_task = task.to_dict()
+        serialized_tool_task["tool"]["module_name"] = "tests.mocks.mock_tool.tool"
+        assert isinstance(serialized_tool_task, dict)
+
+        deserialized_tool_task = ToolTask.from_dict(serialized_tool_task)
+        assert isinstance(deserialized_tool_task, ToolTask)
```
Comment (collindutter) [inline tests/unit/tasks/test_tool_task.py:275]: Rename to `test_to_dict` and `test_from_dict`

File: tests/unit/tools/test_base_tool.py [hunk 10]
```diff
@@ -279,3 +280,31 @@ def test_to_native_tool_name(self, tool, mocker):
...
+        tool = MockTool()
+        action = ToolAction(input={}, name="", tag="")
+
+        serialized_tool = tool.to_dict()
+        assert isinstance(serialized_tool, dict)
+
+        deserialized_tool = MockTool.from_dict(serialized_tool)
+        assert isinstance(deserialized_tool, BaseTool)
+
+        assert deserialized_tool.execute(tool.test_list_output, ActionsSubtask("foo"), action).to_text() == "foo\n\nbar"
```
Comment (collindutter) [inline tests/unit/tools/test_base_tool.py:310]: Rename to `test_to_dict` and `test_from_dict`

**Review by collindutter — CHANGES_REQUESTED** [review]

**Review by collindutter — APPROVED** [review]
Great work!

**Pull request merged**